src/v/resource_mgmt/memory_groups.h

```diff
@@ -11,48 +11,73 @@
 
 #pragma once
 
-#include "seastarx.h"
+#include <cstddef>
 
-#include <seastar/core/memory.hh>
+/**
+ * Centralized unit for memory management.
+ *
+ * Works via a share system. First we subtract the amount of memory the user
+ * decides to use for their WebAssembly functions, then we carve out 30% for the
+ * batch cache. The remaining subsystems are distributed memory via a share
+ * system.
+ */
+class system_memory_groups {
+public:
+    system_memory_groups(size_t total_system_memory, bool wasm_enabled);
+
+    size_t kafka_total_memory() const;
 
-// centralized unit for memory management
-struct memory_groups {
-    static size_t kafka_total_memory() {
-        // 30%
-        return ss::memory::stats().total_memory() * .30;
-    }
     /// \brief includes raft & all services
-    static size_t rpc_total_memory() {
-        // 20%
-        return ss::memory::stats().total_memory() * .20;
-    }
+    size_t rpc_total_memory() const;
 
     /**
      * The target allocation size for the chunk cache. This is a soft target,
      * and may be expanded as needed by segment appenders, or reclaimed from by
      * seastar under memory pressure.
-     *
-     * add upper bound of 30 pct of memory as a hard outstanding limit.
      */
-    static size_t chunk_cache_min_memory() {
-        return ss::memory::stats().total_memory() * .10; // NOLINT
-    }
+    size_t chunk_cache_min_memory() const;
 
     /**
      * Upper bound on the amount of outstanding memory for inflight write
      * requests. Requests above this limit will wait for an existing chunk to be
      * returned to the cache.
      */
-    static size_t chunk_cache_max_memory() {
-        return ss::memory::stats().total_memory() * .30; // NOLINT
-    }
+    size_t chunk_cache_max_memory() const;
 
-    static size_t recovery_max_memory() {
-        return ss::memory::stats().total_memory() * .10;
-    }
+    size_t recovery_max_memory() const;
 
     /// Max memory that both cloud storage uploads and read-path could use
-    static size_t tiered_storage_max_memory() {
-        return ss::memory::stats().total_memory() * .10;
-    }
+    size_t tiered_storage_max_memory() const;
+
+    /// Max memory that data transform subsystem should use.
+    size_t data_transforms_max_memory() const;
+
+private:
+    /**
+     * Total memory for a core after the user's Wasm reservation.
+     */
+    size_t total_memory() const;
+    /**
+     * The amount of memory we reserve for the rest of the system to use, after
+     * we take the minimum amount required for the batch cache.
+     */
+    template<size_t shares>
+    size_t subsystem_memory() const;
+
+    size_t _total_system_memory;
+    bool _wasm_enabled;
 };
+
+/**
+ * Initialize the memory groups for **this core**.
+ *
+ * Initializing memory groups reads from cluster configuration values that can
+ * change, but require a restart to take effect, so we read and lock in the
```

Comment:
If you change a restart-needed property in the config system from x to y, does the config system start returning y after that point in the current process even before restart?

That seems like it would be a pretty big bug in the config system since it would be returning the new value to subsystems which aren't prepared to handle it.

Reply:
> That seems like it would be a pretty big bug in the config system since it would be returning the new value to subsystems which aren't prepared to handle it.

Based on my reading of the config manager I think this is the case because needs_restart::yes assumes that you're only reading these values on startup (which AFAIK isn't true here).

Let me verify that hypothesis and get back to this comment.

Reply:
I've verified this is the case:

For example take enable_transactions which requires a config change to enable. See here where it is read dynamically: 

https://github.com/redpanda-data/redpanda/blob/9db503d95bcd60249f58dbafe57bbcc8573bf49c/src/v/redpanda/admin_server.cc#L3617

It's on by default and run the following:

```
$ rpk cluster config set enable_transactions false
Successfully updated configuration. New configuration version is 4.

Cluster needs to be restarted. See more details with 'rpk cluster config status'.
$ rpk cluster config get enable_transactions
false

$ curl http://localhost:9644/v1/transactions
{"message": "Transaction are disabled", "code": 400}%  
```

Reply:
Posted in slack about this: https://redpandadata.slack.com/archives/C044RD18NMV/p1698694579905929